### ISSUE TITLE
Time Series: Change legend order

### ISSUE THREAD

**Issue by szerwi** [grafana/grafana#30907]
**What would you like to be added**:
It would be nice to have option to change queries order in Time Series panel's legend.
Currently it can be changed only by changing queries order. Unfortunately, changing queries order is also changing rendering order. If there are 2 queries with the same value, one is being covered by the other one.
Currently, the graph is being rendered from first to last query, so the query which is first in the legend will get covered by the query which is, for example, third in the legend.
![image](https://user-images.githubusercontent.com/19311511/106913438-4760fa00-6704-11eb-9b5d-378ffc01a0b9.png)

**Why is this needed**:
In my view, first value in the legend should be the most important on the graph and it shouldn't get covered by second or third value.
It could be changed permanently (reverse legend order), but the best way would be to add option to manually set queries order in legend.

**Comment by hugohaggmark** [comment]
Possible related: #6932 & #18186

**Comment by torkelo** [comment]
This works the same as the old graph, the series are rendered in same order as the legend. Don't think we can/should change that

**Comment by leeoniya** [comment]
isn't there a transform that can re-order fields?

also somewhat related: https://github.com/leeoniya/uPlot/issues/386 (zIndex for series)

**Comment by szerwi** [comment]
> This works the same as the old graph, the series are rendered in same order as the legend. Don't think we can/should change that

In old graph panel there is a possibility to sort legend using the parameters. Changing the legend's order does not change the rendering order, so I guess similar mechanism could be used to change the order manually.
BTW, new time series panel does not have the ability to sort the legend using the parameter, only old graph panel has it.

**Comment by dprokop** [comment]
@szerwi indeed the time series panel does not have legend sorting implemented. Added it to the backlog for time series/graph work: https://github.com/grafana/grafana/issues/30564

**Comment by Yuriy-Kuchinskiy** [comment]
@dprokop it looks like #30564 got resolved/closed by you, with comment that remaining open tasks will be tracked separately. Should this issue be reopened then or there is a need to create new one? I'm using grafana 8.2.2 and the issue is still there...

see this conversation as well:
[Unable to sort table legend on name or values](https://community.grafana.com/t/unable-to-sort-table-legend-on-name-or-values/48847)

please advice on the next step?

**Comment by btorsaldivar** [comment]
> @dprokop it looks like #30564 got resolved/closed by you, with comment that remaining open tasks will be tracked separately. Should this issue be reopened then or there is a need to create new one? I'm using grafana 8.2.2 and the issue is still there...
> 
> see this conversation as well: [Unable to sort table legend on name or values](https://community.grafana.com/t/unable-to-sort-table-legend-on-name-or-values/48847)
> 
> please advice on the next step?

Adding my vote for this very important feature which exist in the "old" graph.

**Comment by data-dude** [comment]
This is still problem with 8.2

**Comment by jgitlin-p21** [comment]
> @szerwi indeed the time series panel does not have legend sorting implemented. Added it to the backlog for time series/graph work: #30564

@dprokop #30564 is now closed with your comment:

> This one has been with us for a long time and majority of the items here are already tracked as independent issues/discussions

But _this_ issue ("Time Series: Change legend order") is closed because it's being tracked in #30564

Can we reopen this please? Thank you!

**Comment by luisgf** [comment]
Yes, even in 8.2.3 the sorting problem persists.

**Comment by zuchka** [comment]
https://github.com/grafana/grafana/pull/40226

should solve this in 8.3

**Comment by szerwi** [comment]
@zuchka I think it is not the same. #40226 adds ability to sort the legend, but you still can't manually change the order of series.

**Comment by derekperkins** [comment]
@zuchka Is there any feedback whether this will be reopened or addressed separately?

**Comment by zuchka** [comment]
hey @derekperkins et al.,

because this is a feature request, I think the most effective thing we can do to encourage implementation is to upvote this Discussion, which references this issue and others:

https://github.com/grafana/grafana/discussions/41104

**Comment by mr-manuel** [comment]
If the PR is closed and this should already be implemented, someone know how to sort the legend by alias?

I'm using Grafana 8.4.3 but found no way to do this.

![grafik](https://user-images.githubusercontent.com/12592765/156943704-eb73118e-bda8-4daa-bb31-7f192f84a8fe.png)

**Comment by roykrikke** [comment]
Same here, I also don't know how ti sort the legend.

**Comment by bluefishforsale** [comment]
Does anyone have a work-around?

**Comment by hippocritical** [comment]
still waiting for a fix, imo this is a major oversight to not have it sortable by name, only by properties ...

**Comment by data-dude** [comment]
Sort is working for me in Grafana 9.1

**Comment by mountrcg** [comment]
@data-dude could you describe how?

**Comment by nidhi-naviranj** [comment]
> @data-dude could you describe how?

You need to add in the `legend` configuration of the timeseries panel
 `"sortBy": "<Calculation method you choose>",
   "sortDesc": <bool>`

**Comment by samloik** [comment]
You need those Legend - Mode - Table!!!

**Comment by slnj** [comment]
@data-dude could you add a screenshot where to set this?
I'm using version 9.4.3 and cannot find where to set this in legend configuration of the time series panel
Thanks

**Comment by data-dude** [comment]
@slnj Under Tooltip I see Values sort order and I select Descending

**Comment by szerwi** [comment]
@data-dude @nidhi-naviranj This allows only to sort the legend by values. You still cannot sort by names or choose the legend order by yourself.

**Comment by nidhi-naviranj** [comment]
> @data-dude @nidhi-naviranj This allows only to sort the legend by values. You still cannot sort by names or choose the legend order by yourself.

You can, once you have legend. Click on the legend value on UI. First click will sort it out by ascending, next will be descending and the third one would be by alphabetical order. (P.S the order can be different but it's three orders, so you can hit and trial to get the order for your metrics)

**Comment by gitneep** [comment]
When you have a stacked graph sorting is important, so that for example the biggest shares come at the bottom with the smaller ones on top, which makes it a lot more readable and comparable between graphs that work on the same data. Sorting the legend values themselves works but the graph itself doesn't follow, this was working in the old time series graph that I had in Grafana 7 and earlier. Can this functionality be brought back please?

**Comment by gitneep** [comment]
This feels like such a basic function, both the stacked graph itself to be able to be sorted as well as the legend values. Is anyone working on this?

**Comment by nmarrs** [comment]
Hi - the DataViz squad is planning on improving our legends across all visualization types this quarter. I've [created this issue to track this feature request](https://github.com/grafana/grafana/issues/73442) 👍

**Comment by mpeter50** [comment]
It is such a shame that the issue has been closed without doing anything with it. This would be a very useful feature, at least for stacked charts, but possibly for others too. Let me show below why.

I think I have put together a workaround for prometheus data sources, but it may also be doable with some of the others.
It is not complicated, but I'm afraid it is quite wasteful with the server's (CPU) resources.
Unfortunately the "Organize fields" transformation did not want to accept my data.

Maybe you have noticed that the lines are sorted alphabetically by their labels.
By renaming them (with label replace) to start with a number, you can basically force an ordering for the lines.

This is how the chart has looked before renaming the labels:

![image](https://github.com/grafana/grafana/assets/83356418/22d62419-2581-4006-a617-28de8c7a5f98)

After renaming the labels like this:

![image](https://github.com/grafana/grafana/assets/83356418/56454c0a-07a8-49f3-b56f-1e604bc6a998)

Now the chart looks like this:

![image](https://github.com/grafana/grafana/assets/83356418/945b93a5-6e99-4335-80cb-64f29c552999)

Basically what I have done is I made those that usually change less to be the lower layers, and those that change more to be higher up.

Much, much more readable, right?

I know it wont be useful for everyone as it is, but I hope that it may serve as an inspiration for something that suits your case better.

**Comment by leeoniya** [comment]
> Unfortunately the "Organize fields" transformation did not want to accept my data.

organize feilds requires a simgle frame. for prometheus you can simply use the outer join on time transformation or prepare time series transformation. then you can use organize fields.
